pkg/sql/row/deleter.go

```diff
@@ -212,35 +214,17 @@ func (rd *Deleter) encodeValueForPrimaryIndexFamily(
 	}
 
 	rd.rawValueBuf = rd.rawValueBuf[:0]
-	var lastColID descpb.ColumnID
 	familySortedColumnIDs, ok := rd.Helper.SortedColumnFamily(family.ID)
 	if !ok {
 		return roachpb.Value{}, errors.AssertionFailedf("invalid family sorted column id map")
 	}
-	for _, colID := range familySortedColumnIDs {
-		idx, ok := rd.FetchColIDtoRowIndex.Get(colID)
-		if !ok || values[idx] == tree.DNull {
-			continue
-		}
-
-		if skip := rd.Helper.SkipColumnNotInPrimaryIndexValue(colID, values[idx]); skip {
-			continue
-		}
-
-		col := rd.FetchCols[idx]
-		if lastColID > col.GetID() {
-			return roachpb.Value{}, errors.AssertionFailedf("cannot write column id %d after %d", col.GetID(), lastColID)
-		}
-		colIDDelta := valueside.MakeColumnIDDelta(lastColID, col.GetID())
-		lastColID = col.GetID()
-		var err error
-		rd.rawValueBuf, err = valueside.Encode(rd.rawValueBuf, colIDDelta, values[idx], nil)
-		if err != nil {
-			return roachpb.Value{}, err
-		}
+	var err error
+	rd.rawValueBuf, err = rd.Helper.encodePrimaryIndexValuesToBuf(values, rd.FetchColIDtoRowIndex, familySortedColumnIDs, rd.FetchCols, rd.rawValueBuf)
+	if err != nil {
+		return roachpb.Value{}, err
 	}
 	ret := roachpb.Value{}
-	if len(rd.rawValueBuf) > 0 {
+	if family.ID == 0 || len(rd.rawValueBuf) > 0 {
```

Comment:
This is an area where we've made it a little hard to test because we don't support column families in a lot of our tooling. But here is my understanding. Consider the case where all of the not-null columns in the primary index are contained in the prety key and there are not composite-encoded columns in the primary key. In this case, the key contains all of the data we want to write. Here, there are two cases:

1. For family 0, our expected value should still contain a non-empty value. If it didn't it would look like a tombstone. So, we put in an empty tuple.

```
> select crdb_internal.pretty_key(key, -1), crdb_internal.pretty_value(value) from                                                            
                                -> crdb_internal.scan(crdb_internal.table_span('foo'::regclass::oid::int));                                                                    
  crdb_internal.pretty_key  | crdb_internal.pretty_value
----------------------------+-----------------------------
  /Tenant/2/Table/112/1/1/0 | /TUPLE/
```

2. For other column families, if there are no values to store into the family, then we don't expect a key to exist at all, so our expectation should be null.

Reply:
every day i wake up grateful that i'm not a sql engineer.

Reply:
😛